testserver: broken with latest build

ts.WaitForInit(db) blocks indefinitely on my system (Ubuntu 16.04). Is it possible that the latest cockroachdb build breaks the testserver package?

I added a manual break after 5 seconds of iterating and logged the error and stdout/stderr from the testserver;

    func (ts *TestServer) WaitForInit(db *sql.DB) {
		cur, max := 0, 500
		for {
			// We issue a query that fails both on connection errors and on the
			// system database not existing.
			if _, err := db.Query("SHOW DATABASES"); err == nil {
				return
			} else if cur > max {
				log.Println(ts.Stdout())
				log.Println(ts.Stderr())
				panic(err)
			}
			cur++
			time.Sleep(time.Millisecond * 10)
		}
	}

Output (cleaned up):

    2016/12/04 15:52:31 downloading https://s3.amazonaws.com/cockroach/cockroach/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706 to /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706, this may take some time
    2016/12/04 15:52:44 Using automatically-downloaded binary: /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706
    2016/12/04 15:52:44 process 15520 started: /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706 start --logtostderr --insecure --port=0 --http-port=0 --socket=/tmp/cockroach-testserver200397412/.s.PGSQL.26257 --store=/tmp/cockroach-testserver200397412
    waiting for database to initialize
    2016/12/04 15:52:50 CockroachDB node starting at 2016-12-04 15:52:44.767370201 +0100 CET

    --- FAIL: TestRunServer (19.29s)
    panic: pq: database "information_schema" does not exist [recovered]
        panic: pq: database "information_schema" does not exist

Testserver Stdout:

    build:      9111f75 @ 2016/12/04 13:07:50 (go1.7.3)
    admin:      http://m47:45529
    sql:        postgresql://root@m47:40251?sslmode=disable
    socket:     /tmp/cockroach-testserver200397412/.s.PGSQL.26257
    logs:       /tmp/cockroach-testserver200397412/logs
    store[0]:   path=/tmp/cockroach-testserver200397412
    status:     initialized new cluster
    clusterID:  fcbd6d02-c65b-4930-be7f-6859769ab779
    nodeID:     1

Testserver Stderr:

    2016/12/04 15:52:50 I161204 15:52:44.622880 1 cli/start.go:318  CockroachDB 9111f75 (linux amd64, built 2016/12/04 13:07:50, go1.7.3)
    I161204 15:52:44.723164 1 cli/start.go:334  starting cockroach node
    I161204 15:52:44.723274 1 cli/start.go:336  using local environment variables: COCKROACH_MAX_OFFSET
    W161204 15:52:44.723330 1 server/server.go:154  [n?] running in insecure mode, this is strongly discouraged. See --insecure.
    I161204 15:52:44.723455 1 gossip/gossip.go:248  [n?] initial resolvers: []
    W161204 15:52:44.723495 1 gossip/gossip.go:1124  [n?] no resolvers found; use --join to specify a connected node
    I161204 15:52:44.725427 1 storage/engine/rocksdb.go:326  opening rocksdb instance at "/tmp/cockroach-testserver200397412"
    I161204 15:52:44.730046 50 kv/dist_sender.go:302  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
    I161204 15:52:44.759296 1 server/config.go:443  1 storage engine initialized
    I161204 15:52:44.759864 1 server/node.go:419  [n?] store [n0,s0] not bootstrapped
    I161204 15:52:44.761700 63 storage/replica_proposal.go:348  [s1,r1/1:/M{in-ax},@c420276900] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 411350h52m53.761170701s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=2016-12-04 14:52:44.761649824 +0000 UTC]
    I161204 15:52:44.762300 1 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
    I161204 15:52:44.762384 1 server/node.go:348  [n?] **** cluster fcbd6d02-c65b-4930-be7f-6859769ab779 has been created
    I161204 15:52:44.762404 1 server/node.go:349  [n?] **** add additional nodes by specifying --join=m47:40251
    I161204 15:52:44.762671 1 base/node_id.go:62  [n1] NodeID set to 1
    I161204 15:52:44.763014 1 storage/store.go:1223  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
    I161204 15:52:44.763051 1 server/node.go:432  [n1] initialized store [n1,s1]: {Capacity:105555197952 Available:61341224960 RangeCount:1 LeaseCount:1}
    I161204 15:52:44.763078 1 server/node.go:317  [n1] node ID 1 initialized
    I161204 15:52:44.763156 1 gossip/gossip.go:290  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"m47:40251" > attrs:<> locality:<> 
    I161204 15:52:44.763341 1 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
    I161204 15:52:44.763500 1 server/node.go:562  [n1] connecting to gossip network to verify cluster ID...
    I161204 15:52:44.763535 1 server/node.go:582  [n1] node connected via gossip and verified as part of cluster "fcbd6d02-c65b-4930-be7f-6859769ab779"
    I161204 15:52:44.763582 1 server/node.go:367  [n1] node=1: started with [[]=/tmp/cockroach-testserver200397412] engine(s) and attributes []
    I161204 15:52:44.763635 1 sql/executor.go:291  [n1] creating distSQLPlanner with address {tcp m47:40251}
    I161204 15:52:44.765237 28 storage/split_queue.go:103  [n1,split,s1,r1/1:/M{in-ax},@c420277200] splitting at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
    I161204 15:52:44.766902 28 storage/replica_command.go:2369  [n1,split,s1,r1/1:/M{in-ax},@c420277200] initiating a split of this range at key /Table/11 [r2]
    I161204 15:52:44.767156 1 server/server.go:633  [n1] starting http server at m47:45529
    I161204 15:52:44.767177 1 server/server.go:634  [n1] starting grpc/postgres server at m47:40251
    I161204 15:52:44.767198 1 server/server.go:635  [n1] advertising CockroachDB node at m47:40251
    I161204 15:52:44.767261 1 server/server.go:637  [n1] starting postgres server at unix:/tmp/cockroach-testserver200397412/.s.PGSQL.26257
    I161204 15:52:44.768282 106 server/updates.go:158  [n1,updates check] No previous updates check time.
    E161204 15:52:44.775774 29 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.775884 28 storage/queue.go:598  [n1,split,s1,r1/1:/{Min-Table/11},@c420277200] unable to split [n1,s1,r1/1:/{Min-Table/11}] at key "/Table/12/0": key range /Table/12/0-/Table/12/0 outside of bounds of range /Min-/Max
    E161204 15:52:44.775933 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.776003 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    I161204 15:52:44.776052 28 storage/split_queue.go:103  [n1,split,s1,r2/1:/{Table/11-Max},@c420262480] splitting at keys [/Table/12/0 /Table/13/0 /Table/14/0]
    I161204 15:52:44.776098 28 storage/replica_command.go:2369  [n1,split,s1,r2/1:/{Table/11-Max},@c420262480] initiating a split of this range at key /Table/12 [r3]
    E161204 15:52:44.780195 29 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.780334 28 storage/queue.go:598  [n1,split,s1,r2/1:/Table/1{1-2},@c420262480] unable to split [n1,s1,r2/1:/Table/1{1-2}] at key "/Table/13/0": key range /Table/13/0-/Table/13/0 outside of bounds of range /Table/11-/Max
    E161204 15:52:44.780492 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.780576 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    I161204 15:52:44.780705 28 storage/split_queue.go:103  [n1,split,s1,r3/1:/{Table/12-Max},@c420262d80] splitting at keys [/Table/13/0 /Table/14/0]
    I161204 15:52:44.780758 28 storage/replica_command.go:2369  [n1,split,s1,r3/1:/{Table/12-Max},@c420262d80] initiating a split of this range at key /Table/13 [r4]
    E161204 15:52:44.784909 29 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.784938 28 storage/queue.go:598  [n1,split,s1,r3/1:/Table/1{2-3},@c420262d80] unable to split [n1,s1,r3/1:/Table/1{2-3}] at key "/Table/14/0": key range /Table/14/0-/Table/14/0 outside of bounds of range /Table/12-/Max
    I161204 15:52:44.785024 28 storage/split_queue.go:103  [n1,split,s1,r4/1:/{Table/13-Max},@c4201ea900] splitting at keys [/Table/14/0]
    I161204 15:52:44.785064 28 storage/replica_command.go:2369  [n1,split,s1,r4/1:/{Table/13-Max},@c4201ea900] initiating a split of this range at key /Table/14 [r5]
    E161204 15:52:44.785120 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.785190 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.785246 110 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789105 29 storage/queue.go:609  [n1,replicate,s1,r4/1:/Table/1{3-4},@c4201ea900] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789255 29 storage/queue.go:609  [n1,replicate,s1,r5/1:/{Table/14-Max},@c42024b680] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789312 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789372 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789426 110 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789477 110 storage/queue.go:609  [n1,replicate,s1,r4/1:/Table/1{3-4},@c4201ea900] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    I161204 15:52:44.811683 134 sql/event_log.go:95  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:m47:40251} Attrs: Locality:} ClusterID:fcbd6d02-c65b-4930-be7f-6859769ab779 StartedAt:1480863164763543657}
    I161204 15:52:45.353809 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160428, details: https://www.cockroachlabs.com/docs/beta-20160428.html
    I161204 15:52:45.353825 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160505, details: https://www.cockroachlabs.com/docs/beta-20160505.html
    I161204 15:52:45.353833 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160512, details: https://www.cockroachlabs.com/docs/beta-20160512.html
    I161204 15:52:45.353843 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160519, details: https://www.cockroachlabs.com/docs/beta-20160519.html
    I161204 15:52:45.353851 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160526, details: https://www.cockroachlabs.com/docs/beta-20160526.html
    I161204 15:52:45.353860 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160602, details: https://www.cockroachlabs.com/docs/beta-20160602.html
    I161204 15:52:45.353868 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160609, details: https://www.cockroachlabs.com/docs/beta-20160609.html
    I161204 15:52:45.353876 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160616, details: https://www.cockroachlabs.com/docs/beta-20160616.html
    I161204 15:52:45.353885 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160714, details: https://www.cockroachlabs.com/docs/beta-20160714.html
    I161204 15:52:45.353893 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160728, details: https://www.cockroachlabs.com/docs/beta-20160728.html
    I161204 15:52:45.353901 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160908, details: https://www.cockroachlabs.com/docs/beta-20160908.html
    I161204 15:52:45.353909 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160915, details: https://www.cockroachlabs.com/docs/beta-20160915.html
    I161204 15:52:45.353921 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160929, details: https://www.cockroachlabs.com/docs/beta-20160929.html
    I161204 15:52:45.353930 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161006, details: https://www.cockroachlabs.com/docs/beta-20161006.html
    I161204 15:52:45.353938 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161013, details: https://www.cockroachlabs.com/docs/beta-20161013.html
    I161204 15:52:45.353946 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161027, details: https://www.cockroachlabs.com/docs/beta-20161027.html
    I161204 15:52:45.353956 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161103, details: https://www.cockroachlabs.com/docs/beta-20161103.html
    I161204 15:52:45.353964 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161201, details: https://www.cockroachlabs.com/docs/beta-20161201.html

The commit you're running was a test-only change; is there a previous ref
that doesn't exhibit this problem?

Also, are you running this server with a data directory created by an
earlier version, or is it fresh?

On Dec 4, 2016 10:06, "boljen" <notifications@github.com> wrote:

> ts.WaitForInit(db) blocks indefinitely on my system (Ubuntu 16.04). Is it
> possible that the latest cockroachdb build breaks the testserver package?
>
> I added a manual break after 5 seconds of iterating and logged the error,
> as well as the stdout/stderr from the testserver;
>
> func (ts *TestServer) WaitForInit(db *sql.DB) {
> 	cur, max := 0, 500
> 	for {
> 		// We issue a query that fails both on connection errors and on the
> 		// system database not existing.
> 		if _, err := db.Query("SHOW DATABASES"); err == nil {
> 			return
> 		} else if cur > max {
> 			log.Println(ts.Stdout())
> 			log.Println(ts.Stderr())
> 			panic(err)
> 		}
> 		cur++
> 		time.Sleep(time.Millisecond * 10)
> 	}
> }
>
> Output (cleaned up):
>
> 2016/12/04 15:52:31 downloading https://s3.amazonaws.com/cockroach/cockroach/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706 to /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706, this may take some time
> 2016/12/04 15:52:44 Using automatically-downloaded binary: /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706
> 2016/12/04 15:52:44 process 15520 started: /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706 start --logtostderr --insecure --port=0 --http-port=0 --socket=/tmp/cockroach-testserver200397412/.s.PGSQL.26257 --store=/tmp/cockroach-testserver200397412
> waiting for database to initialize
> 2016/12/04 15:52:50 CockroachDB node starting at 2016-12-04 15:52:44.767370201 +0100 CET
>
> --- FAIL: TestRunServer (19.29s)
> panic: pq: database "information_schema" does not exist [recovered]
>     panic: pq: database "information_schema" does not exist
>
> Testserver Stdout:
>
> build:      9111f75 @ 2016/12/04 13:07:50 (go1.7.3)
> admin:      http://m47:45529
> sql:        postgresql://root@m47:40251?sslmode=disable
> socket:     /tmp/cockroach-testserver200397412/.s.PGSQL.26257
> logs:       /tmp/cockroach-testserver200397412/logs
> store[0]:   path=/tmp/cockroach-testserver200397412
> status:     initialized new cluster
> clusterID:  fcbd6d02-c65b-4930-be7f-6859769ab779
> nodeID:     1
>
> Testserver Stderr:
>
> 2016/12/04 15:52:50 I161204 15:52:44.622880 1 cli/start.go:318  CockroachDB 9111f75 (linux amd64, built 2016/12/04 13:07:50, go1.7.3)
> I161204 15:52:44.723164 1 cli/start.go:334  starting cockroach node
> I161204 15:52:44.723274 1 cli/start.go:336  using local environment variables: COCKROACH_MAX_OFFSET
> W161204 15:52:44.723330 1 server/server.go:154  [n?] running in insecure mode, this is strongly discouraged. See --insecure.
> I161204 15:52:44.723455 1 gossip/gossip.go:248  [n?] initial resolvers: []
> W161204 15:52:44.723495 1 gossip/gossip.go:1124  [n?] no resolvers found; use --join to specify a connected node
> I161204 15:52:44.725427 1 storage/engine/rocksdb.go:326  opening rocksdb instance at "/tmp/cockroach-testserver200397412"
> I161204 15:52:44.730046 50 kv/dist_sender.go:302  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
> I161204 15:52:44.759296 1 server/config.go:443  1 storage engine initialized
> I161204 15:52:44.759864 1 server/node.go:419  [n?] store [n0,s0] not bootstrapped
> I161204 15:52:44.761700 63 storage/replica_proposal.go:348  [s1,r1/1:/M{in-ax},@c420276900] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 411350h52m53.761170701s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=2016-12-04 14:52:44.761649824 +0000 UTC]
> I161204 15:52:44.762300 1 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
> I161204 15:52:44.762384 1 server/node.go:348  [n?] **** cluster fcbd6d02-c65b-4930-be7f-6859769ab779 has been created
> I161204 15:52:44.762404 1 server/node.go:349  [n?] **** add additional nodes by specifying --join=m47:40251
> I161204 15:52:44.762671 1 base/node_id.go:62  [n1] NodeID set to 1
> I161204 15:52:44.763014 1 storage/store.go:1223  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
> I161204 15:52:44.763051 1 server/node.go:432  [n1] initialized store [n1,s1]: {Capacity:105555197952 Available:61341224960 RangeCount:1 LeaseCount:1}
> I161204 15:52:44.763078 1 server/node.go:317  [n1] node ID 1 initialized
> I161204 15:52:44.763156 1 gossip/gossip.go:290  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"m47:40251" > attrs:<> locality:<>
> I161204 15:52:44.763341 1 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
> I161204 15:52:44.763500 1 server/node.go:562  [n1] connecting to gossip network to verify cluster ID...
> I161204 15:52:44.763535 1 server/node.go:582  [n1] node connected via gossip and verified as part of cluster "fcbd6d02-c65b-4930-be7f-6859769ab779"
> I161204 15:52:44.763582 1 server/node.go:367  [n1] node=1: started with [[]=/tmp/cockroach-testserver200397412] engine(s) and attributes []
> I161204 15:52:44.763635 1 sql/executor.go:291  [n1] creating distSQLPlanner with address {tcp m47:40251}
> I161204 15:52:44.765237 28 storage/split_queue.go:103  [n1,split,s1,r1/1:/M{in-ax},@c420277200] splitting at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
> I161204 15:52:44.766902 28 storage/replica_command.go:2369  [n1,split,s1,r1/1:/M{in-ax},@c420277200] initiating a split of this range at key /Table/11 [r2]
> I161204 15:52:44.767156 1 server/server.go:633  [n1] starting http server at m47:45529
> I161204 15:52:44.767177 1 server/server.go:634  [n1] starting grpc/postgres server at m47:40251
> I161204 15:52:44.767198 1 server/server.go:635  [n1] advertising CockroachDB node at m47:40251
> I161204 15:52:44.767261 1 server/server.go:637  [n1] starting postgres server at unix:/tmp/cockroach-testserver200397412/.s.PGSQL.26257
> I161204 15:52:44.768282 106 server/updates.go:158  [n1,updates check] No previous updates check time.
> E161204 15:52:44.775774 29 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.775884 28 storage/queue.go:598  [n1,split,s1,r1/1:/{Min-Table/11},@c420277200] unable to split [n1,s1,r1/1:/{Min-Table/11}] at key "/Table/12/0": key range /Table/12/0-/Table/12/0 outside of bounds of range /Min-/Max
> E161204 15:52:44.775933 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.776003 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> I161204 15:52:44.776052 28 storage/split_queue.go:103  [n1,split,s1,r2/1:/{Table/11-Max},@c420262480] splitting at keys [/Table/12/0 /Table/13/0 /Table/14/0]
> I161204 15:52:44.776098 28 storage/replica_command.go:2369  [n1,split,s1,r2/1:/{Table/11-Max},@c420262480] initiating a split of this range at key /Table/12 [r3]
> E161204 15:52:44.780195 29 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.780334 28 storage/queue.go:598  [n1,split,s1,r2/1:/Table/1{1-2},@c420262480] unable to split [n1,s1,r2/1:/Table/1{1-2}] at key "/Table/13/0": key range /Table/13/0-/Table/13/0 outside of bounds of range /Table/11-/Max
> E161204 15:52:44.780492 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.780576 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> I161204 15:52:44.780705 28 storage/split_queue.go:103  [n1,split,s1,r3/1:/{Table/12-Max},@c420262d80] splitting at keys [/Table/13/0 /Table/14/0]
> I161204 15:52:44.780758 28 storage/replica_command.go:2369  [n1,split,s1,r3/1:/{Table/12-Max},@c420262d80] initiating a split of this range at key /Table/13 [r4]
> E161204 15:52:44.784909 29 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.784938 28 storage/queue.go:598  [n1,split,s1,r3/1:/Table/1{2-3},@c420262d80] unable to split [n1,s1,r3/1:/Table/1{2-3}] at key "/Table/14/0": key range /Table/14/0-/Table/14/0 outside of bounds of range /Table/12-/Max
> I161204 15:52:44.785024 28 storage/split_queue.go:103  [n1,split,s1,r4/1:/{Table/13-Max},@c4201ea900] splitting at keys [/Table/14/0]
> I161204 15:52:44.785064 28 storage/replica_command.go:2369  [n1,split,s1,r4/1:/{Table/13-Max},@c4201ea900] initiating a split of this range at key /Table/14 [r5]
> E161204 15:52:44.785120 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.785190 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.785246 110 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.789105 29 storage/queue.go:609  [n1,replicate,s1,r4/1:/Table/1{3-4},@c4201ea900] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.789255 29 storage/queue.go:609  [n1,replicate,s1,r5/1:/{Table/14-Max},@c42024b680] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.789312 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.789372 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.789426 110 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> E161204 15:52:44.789477 110 storage/queue.go:609  [n1,replicate,s1,r4/1:/Table/1{3-4},@c4201ea900] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
> I161204 15:52:44.811683 134 sql/event_log.go:95  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:m47:40251} Attrs: Locality:} ClusterID:fcbd6d02-c65b-4930-be7f-6859769ab779 StartedAt:1480863164763543657}
> I161204 15:52:45.353809 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160428, details: https://www.cockroachlabs.com/docs/beta-20160428.html
> I161204 15:52:45.353825 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160505, details: https://www.cockroachlabs.com/docs/beta-20160505.html
> I161204 15:52:45.353833 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160512, details: https://www.cockroachlabs.com/docs/beta-20160512.html
> I161204 15:52:45.353843 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160519, details: https://www.cockroachlabs.com/docs/beta-20160519.html
> I161204 15:52:45.353851 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160526, details: https://www.cockroachlabs.com/docs/beta-20160526.html
> I161204 15:52:45.353860 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160602, details: https://www.cockroachlabs.com/docs/beta-20160602.html
> I161204 15:52:45.353868 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160609, details: https://www.cockroachlabs.com/docs/beta-20160609.html
> I161204 15:52:45.353876 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160616, details: https://www.cockroachlabs.com/docs/beta-20160616.html
> I161204 15:52:45.353885 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160714, details: https://www.cockroachlabs.com/docs/beta-20160714.html
> I161204 15:52:45.353893 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160728, details: https://www.cockroachlabs.com/docs/beta-20160728.html
> I161204 15:52:45.353901 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160908, details: https://www.cockroachlabs.com/docs/beta-20160908.html
> I161204 15:52:45.353909 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160915, details: https://www.cockroachlabs.com/docs/beta-20160915.html
> I161204 15:52:45.353921 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160929, details: https://www.cockroachlabs.com/docs/beta-20160929.html
> I161204 15:52:45.353930 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161006, details: https://www.cockroachlabs.com/docs/beta-20161006.html
> I161204 15:52:45.353938 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161013, details: https://www.cockroachlabs.com/docs/beta-20161013.html
> I161204 15:52:45.353946 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161027, details: https://www.cockroachlabs.com/docs/beta-20161027.html
> I161204 15:52:45.353956 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161103, details: https://www.cockroachlabs.com/docs/beta-20161103.html
> I161204 15:52:45.353964 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161201, details: https://www.cockroachlabs.com/docs/beta-20161201.html
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach-go/issues/20>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPEV6nkqcbVaqu3CrBokyHCN4SixCks5rEtcEgaJpZM4LDmxs>
> .
>

Not a duplicate; that is issue was closed because it was created by a bug
in our automation. This issue still seems legitimate to me.

On Dec 4, 2016 10:23, "Tamir Duberstein" <tamir@cockroachlabs.com> wrote:

> The commit you're running was a test-only change; is there a previous ref
> that doesn't exhibit this problem?
>
> Also, are you running this server with a data directory created by an
> earlier version, or is it fresh?
>
> On Dec 4, 2016 10:06, "boljen" <notifications@github.com> wrote:
>
>> ts.WaitForInit(db) blocks indefinitely on my system (Ubuntu 16.04). Is it
>> possible that the latest cockroachdb build breaks the testserver package?
>>
>> I added a manual break after 5 seconds of iterating and logged the error,
>> as well as the stdout/stderr from the testserver;
>>
>> func (ts *TestServer) WaitForInit(db *sql.DB) {
>> 	cur, max := 0, 500
>> 	for {
>> 		// We issue a query that fails both on connection errors and on the
>> 		// system database not existing.
>> 		if _, err := db.Query("SHOW DATABASES"); err == nil {
>> 			return
>> 		} else if cur > max {
>> 			log.Println(ts.Stdout())
>> 			log.Println(ts.Stderr())
>> 			panic(err)
>> 		}
>> 		cur++
>> 		time.Sleep(time.Millisecond * 10)
>> 	}
>> }
>>
>> Output (cleaned up):
>>
>> 2016/12/04 15:52:31 downloading https://s3.amazonaws.com/cockroach/cockroach/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706 to /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706, this may take some time
>> 2016/12/04 15:52:44 Using automatically-downloaded binary: /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706
>> 2016/12/04 15:52:44 process 15520 started: /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706 start --logtostderr --insecure --port=0 --http-port=0 --socket=/tmp/cockroach-testserver200397412/.s.PGSQL.26257 --store=/tmp/cockroach-testserver200397412
>> waiting for database to initialize
>> 2016/12/04 15:52:50 CockroachDB node starting at 2016-12-04 15:52:44.767370201 +0100 CET
>>
>> --- FAIL: TestRunServer (19.29s)
>> panic: pq: database "information_schema" does not exist [recovered]
>>     panic: pq: database "information_schema" does not exist
>>
>> Testserver Stdout:
>>
>> build:      9111f75 @ 2016/12/04 13:07:50 (go1.7.3)
>> admin:      http://m47:45529
>> sql:        postgresql://root@m47:40251?sslmode=disable
>> socket:     /tmp/cockroach-testserver200397412/.s.PGSQL.26257
>> logs:       /tmp/cockroach-testserver200397412/logs
>> store[0]:   path=/tmp/cockroach-testserver200397412
>> status:     initialized new cluster
>> clusterID:  fcbd6d02-c65b-4930-be7f-6859769ab779
>> nodeID:     1
>>
>> Testserver Stderr:
>>
>> 2016/12/04 15:52:50 I161204 15:52:44.622880 1 cli/start.go:318  CockroachDB 9111f75 (linux amd64, built 2016/12/04 13:07:50, go1.7.3)
>> I161204 15:52:44.723164 1 cli/start.go:334  starting cockroach node
>> I161204 15:52:44.723274 1 cli/start.go:336  using local environment variables: COCKROACH_MAX_OFFSET
>> W161204 15:52:44.723330 1 server/server.go:154  [n?] running in insecure mode, this is strongly discouraged. See --insecure.
>> I161204 15:52:44.723455 1 gossip/gossip.go:248  [n?] initial resolvers: []
>> W161204 15:52:44.723495 1 gossip/gossip.go:1124  [n?] no resolvers found; use --join to specify a connected node
>> I161204 15:52:44.725427 1 storage/engine/rocksdb.go:326  opening rocksdb instance at "/tmp/cockroach-testserver200397412"
>> I161204 15:52:44.730046 50 kv/dist_sender.go:302  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
>> I161204 15:52:44.759296 1 server/config.go:443  1 storage engine initialized
>> I161204 15:52:44.759864 1 server/node.go:419  [n?] store [n0,s0] not bootstrapped
>> I161204 15:52:44.761700 63 storage/replica_proposal.go:348  [s1,r1/1:/M{in-ax},@c420276900] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 411350h52m53.761170701s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=2016-12-04 14:52:44.761649824 +0000 UTC]
>> I161204 15:52:44.762300 1 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
>> I161204 15:52:44.762384 1 server/node.go:348  [n?] **** cluster fcbd6d02-c65b-4930-be7f-6859769ab779 has been created
>> I161204 15:52:44.762404 1 server/node.go:349  [n?] **** add additional nodes by specifying --join=m47:40251
>> I161204 15:52:44.762671 1 base/node_id.go:62  [n1] NodeID set to 1
>> I161204 15:52:44.763014 1 storage/store.go:1223  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
>> I161204 15:52:44.763051 1 server/node.go:432  [n1] initialized store [n1,s1]: {Capacity:105555197952 Available:61341224960 RangeCount:1 LeaseCount:1}
>> I161204 15:52:44.763078 1 server/node.go:317  [n1] node ID 1 initialized
>> I161204 15:52:44.763156 1 gossip/gossip.go:290  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"m47:40251" > attrs:<> locality:<>
>> I161204 15:52:44.763341 1 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
>> I161204 15:52:44.763500 1 server/node.go:562  [n1] connecting to gossip network to verify cluster ID...
>> I161204 15:52:44.763535 1 server/node.go:582  [n1] node connected via gossip and verified as part of cluster "fcbd6d02-c65b-4930-be7f-6859769ab779"
>> I161204 15:52:44.763582 1 server/node.go:367  [n1] node=1: started with [[]=/tmp/cockroach-testserver200397412] engine(s) and attributes []
>> I161204 15:52:44.763635 1 sql/executor.go:291  [n1] creating distSQLPlanner with address {tcp m47:40251}
>> I161204 15:52:44.765237 28 storage/split_queue.go:103  [n1,split,s1,r1/1:/M{in-ax},@c420277200] splitting at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
>> I161204 15:52:44.766902 28 storage/replica_command.go:2369  [n1,split,s1,r1/1:/M{in-ax},@c420277200] initiating a split of this range at key /Table/11 [r2]
>> I161204 15:52:44.767156 1 server/server.go:633  [n1] starting http server at m47:45529
>> I161204 15:52:44.767177 1 server/server.go:634  [n1] starting grpc/postgres server at m47:40251
>> I161204 15:52:44.767198 1 server/server.go:635  [n1] advertising CockroachDB node at m47:40251
>> I161204 15:52:44.767261 1 server/server.go:637  [n1] starting postgres server at unix:/tmp/cockroach-testserver200397412/.s.PGSQL.26257
>> I161204 15:52:44.768282 106 server/updates.go:158  [n1,updates check] No previous updates check time.
>> E161204 15:52:44.775774 29 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.775884 28 storage/queue.go:598  [n1,split,s1,r1/1:/{Min-Table/11},@c420277200] unable to split [n1,s1,r1/1:/{Min-Table/11}] at key "/Table/12/0": key range /Table/12/0-/Table/12/0 outside of bounds of range /Min-/Max
>> E161204 15:52:44.775933 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.776003 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> I161204 15:52:44.776052 28 storage/split_queue.go:103  [n1,split,s1,r2/1:/{Table/11-Max},@c420262480] splitting at keys [/Table/12/0 /Table/13/0 /Table/14/0]
>> I161204 15:52:44.776098 28 storage/replica_command.go:2369  [n1,split,s1,r2/1:/{Table/11-Max},@c420262480] initiating a split of this range at key /Table/12 [r3]
>> E161204 15:52:44.780195 29 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.780334 28 storage/queue.go:598  [n1,split,s1,r2/1:/Table/1{1-2},@c420262480] unable to split [n1,s1,r2/1:/Table/1{1-2}] at key "/Table/13/0": key range /Table/13/0-/Table/13/0 outside of bounds of range /Table/11-/Max
>> E161204 15:52:44.780492 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.780576 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> I161204 15:52:44.780705 28 storage/split_queue.go:103  [n1,split,s1,r3/1:/{Table/12-Max},@c420262d80] splitting at keys [/Table/13/0 /Table/14/0]
>> I161204 15:52:44.780758 28 storage/replica_command.go:2369  [n1,split,s1,r3/1:/{Table/12-Max},@c420262d80] initiating a split of this range at key /Table/13 [r4]
>> E161204 15:52:44.784909 29 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.784938 28 storage/queue.go:598  [n1,split,s1,r3/1:/Table/1{2-3},@c420262d80] unable to split [n1,s1,r3/1:/Table/1{2-3}] at key "/Table/14/0": key range /Table/14/0-/Table/14/0 outside of bounds of range /Table/12-/Max
>> I161204 15:52:44.785024 28 storage/split_queue.go:103  [n1,split,s1,r4/1:/{Table/13-Max},@c4201ea900] splitting at keys [/Table/14/0]
>> I161204 15:52:44.785064 28 storage/replica_command.go:2369  [n1,split,s1,r4/1:/{Table/13-Max},@c4201ea900] initiating a split of this range at key /Table/14 [r5]
>> E161204 15:52:44.785120 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.785190 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.785246 110 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.789105 29 storage/queue.go:609  [n1,replicate,s1,r4/1:/Table/1{3-4},@c4201ea900] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.789255 29 storage/queue.go:609  [n1,replicate,s1,r5/1:/{Table/14-Max},@c42024b680] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.789312 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.789372 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.789426 110 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> E161204 15:52:44.789477 110 storage/queue.go:609  [n1,replicate,s1,r4/1:/Table/1{3-4},@c4201ea900] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
>> I161204 15:52:44.811683 134 sql/event_log.go:95  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:m47:40251} Attrs: Locality:} ClusterID:fcbd6d02-c65b-4930-be7f-6859769ab779 StartedAt:1480863164763543657}
>> I161204 15:52:45.353809 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160428, details: https://www.cockroachlabs.com/docs/beta-20160428.html
>> I161204 15:52:45.353825 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160505, details: https://www.cockroachlabs.com/docs/beta-20160505.html
>> I161204 15:52:45.353833 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160512, details: https://www.cockroachlabs.com/docs/beta-20160512.html
>> I161204 15:52:45.353843 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160519, details: https://www.cockroachlabs.com/docs/beta-20160519.html
>> I161204 15:52:45.353851 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160526, details: https://www.cockroachlabs.com/docs/beta-20160526.html
>> I161204 15:52:45.353860 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160602, details: https://www.cockroachlabs.com/docs/beta-20160602.html
>> I161204 15:52:45.353868 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160609, details: https://www.cockroachlabs.com/docs/beta-20160609.html
>> I161204 15:52:45.353876 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160616, details: https://www.cockroachlabs.com/docs/beta-20160616.html
>> I161204 15:52:45.353885 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160714, details: https://www.cockroachlabs.com/docs/beta-20160714.html
>> I161204 15:52:45.353893 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160728, details: https://www.cockroachlabs.com/docs/beta-20160728.html
>> I161204 15:52:45.353901 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160908, details: https://www.cockroachlabs.com/docs/beta-20160908.html
>> I161204 15:52:45.353909 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160915, details: https://www.cockroachlabs.com/docs/beta-20160915.html
>> I161204 15:52:45.353921 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160929, details: https://www.cockroachlabs.com/docs/beta-20160929.html
>> I161204 15:52:45.353930 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161006, details: https://www.cockroachlabs.com/docs/beta-20161006.html
>> I161204 15:52:45.353938 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161013, details: https://www.cockroachlabs.com/docs/beta-20161013.html
>> I161204 15:52:45.353946 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161027, details: https://www.cockroachlabs.com/docs/beta-20161027.html
>> I161204 15:52:45.353956 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161103, details: https://www.cockroachlabs.com/docs/beta-20161103.html
>> I161204 15:52:45.353964 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161201, details: https://www.cockroachlabs.com/docs/beta-20161201.html
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cockroachdb/cockroach-go/issues/20>, or mute the
>> thread
>> <https://github.com/notifications/unsubscribe-auth/ABdsPEV6nkqcbVaqu3CrBokyHCN4SixCks5rEtcEgaJpZM4LDmxs>
>> .
>>
>

--- a/testserver/testserver.go
+++ b/testserver/testserver.go
@@ -57,6 +57,7 @@ package testserver
 import (
 	"database/sql"
 	"errors"
+	"flag"
 	"fmt"
 	"io/ioutil"
 	"log"
@@ -73,7 +74,10 @@ import (
 	"time"
 )
 
-var sqlURLRegexp = regexp.MustCompile("sql:\\s+(postgresql:.+)\n")
+var (
+	sqlURLRegexp = regexp.MustCompile("sql:\\s+(postgresql:.+)\n")
+	customBinary = flag.String("cockroach-binary", "", "Use specified cockroach binary")
+)
 
 const (
 	stateNew     = iota
@@ -136,7 +140,9 @@ func NewDBForTestWithDatabase(t *testing.T, database string) (*sql.DB, func()) {
 		t.Fatal(err)
 	}
 
-	ts.WaitForInit(db)
+	if err := ts.WaitForInit(db); err != nil {
+		t.Fatal(err)
+	}
 
 	return db, func() {
 		_ = db.Close()
@@ -149,12 +155,16 @@ func NewDBForTestWithDatabase(t *testing.T, database string) (*sql.DB, func()) {
 // If the download fails, we attempt just call "cockroach", hoping it is
 // found in your path.
 func NewTestServer() (*TestServer, error) {
-	cockroachBinary, err := downloadLatestBinary()
-	if err == nil {
-		log.Printf("Using automatically-downloaded binary: %s", cockroachBinary)
-	} else {
-		log.Printf("Attempting to use cockroach binary from your PATH")
+	var cockroachBinary string
+	var err error
+	if len(*customBinary) > 0 {
+		cockroachBinary = *customBinary
+		log.Printf("Using custom cockroach binary: %s", cockroachBinary)
+	} else if cockroachBinary, err = downloadLatestBinary(); err != nil {
+		log.Printf("Failed to fetch latest binary: %s, attempting to use cockroach binary from your PATH", err)
 		cockroachBinary = "cockroach"
+	} else {
+		log.Printf("Using automatically-downloaded binary: %s", cockroachBinary)
 	}
 
 	// Force "/tmp/" so avoid OSX's really long temp directory names
@@ -225,19 +235,17 @@ func (ts *TestServer) PGURL() *url.URL {
 	}
 }
 
-// WaitForInit repeatedly looks up the list of databases until
-// the "system" database exists. It ignores all errors as we are
-// waiting for the process to start and complete initialization.
-// This does not timeout, relying instead on test timeouts.
-func (ts *TestServer) WaitForInit(db *sql.DB) {
-	for {
-		// We issue a query that fails both on connection errors and on the
-		// system database not existing.
-		if _, err := db.Query("SHOW DATABASES"); err == nil {
-			return
+// WaitForInit retries until a connection is successfully established.
+func (ts *TestServer) WaitForInit(db *sql.DB) error {
+	var err error
+	for i := 0; i < 50; i++ {
+		if _, err = db.Query("SHOW DATABASES"); err == nil {
+			return err
 		}
-		time.Sleep(time.Millisecond * 10)
+		log.Printf("WaitForInit: %v", err)
+		time.Sleep(time.Millisecond * 100)
 	}
+	return err
 }
 
 // Start runs the process, returning an error on any problems,

